### ISSUE TITLE
ForkAndPullRepo does not work with private repositories

### ISSUE THREAD

**Issue by hferentschik** [jenkins-x/jx#5341]
`ForkAndPullRepo` in `helpers.go` does not work with private repositories, since it does not pass credentials when accessing the repository.

```
$ jx boot
...
validating git repository for dev at URL https://github.com/hf-bee-bot/environment-hardy-jx-private-env-dev.git
creating environment dev git repository for URL: https://github.com/hf-bee-bot/environment-hardy-jx-private-env-dev.git to namespace jx
Duplicated Git repository https://github.com/jenkins-x/jenkins-x-boot-config to https://github.com/hf-bee-bot/environment-hardy-jx-private-env-dev
Setting upstream to https://github.com/hf-bee-bot/environment-hardy-jx-private-env-dev
error: creating dev environment for environment-hardy-jx-private-env-dev: forking and pulling https://github.com/hf-bee-bot/environment-hardy-jx-private-env-dev.git: fetching origin master: git output: remote: Repository not found.
fatal: repository 'https://github.com/hf-bee-bot/environment-hardy-jx-private-env-dev.git/' not found: failed to run 'git fetch origin master' command in directory '/Users/hardy/tmp/issue-5186/boot2/boot', output: 'remote: Repository not found.
fatal: repository 'https://github.com/hf-bee-bot/environment-hardy-jx-private-env-dev.git/' not found'
error: failed to interpret pipeline file jenkins-x.yml: failed to run '/bin/sh -c jx step verify env' command in directory '.', output: ''
```

**Comment by hferentschik** [comment]
Relates to #5186

**Comment by hferentschik** [comment]
So issue is that the method in the Gitter interface do not deal with authentication. In some cases this can be worked around, for example as done with the `Push` command in `DuplicateGitRepoFromCommitish` of `helpers.go`:

```go
	userDetails := provider.UserAuth()
		duplicatePushURL, err := gitter.CreatePushURL(duplicateInfo.CloneURL, &userDetails)
		if err != nil {
			return nil, errors.Wrapf(err, "failed to create push URL for %s", duplicateInfo.CloneURL)
		}
		err = gitter.SetRemoteURL(dir, "origin", duplicateInfo.CloneURL)
		if err != nil {
			return nil, errors.Wrapf(err, "failed to set remote url to %s", duplicateInfo.CloneURL)
		}
		err = gitter.Push(dir, duplicatePushURL, true, false, fmt.Sprintf("%s:%s", "HEAD", toBranch))
```

The idea here is to set the plain remote URL on the repo, but when doing the push specify the URL with authentication added. This works, but requires some additional effort. 
One needs to explicitly create the right URLs for executing the command. Also one needs to specify the full URL opposed to specify the remote name which gets bypassed in this case. 

One approach for solving this problem is continue the approach shown above and in this particular case adjust the calls for `FetchBranch`, maybe more.

Another approach could be to make use of gits [credential helpers](https://git-scm.com/docs/gitcredentials), in particular the [cache](https://git-scm.com/docs/git-credential-cache) helper. The cache helper stores the credentials for a specified time in memory. The credentials never hit the disk. The idea is to add something along this line to `Gitter`:

```go
// CacheCredendials caches the credentials for the specified user, host and path for the specified number of seconds.
CacheCredendials(host string, path string, username string, password string, time int) error
// Clears potentially cached credentials for the specified host, path and user, if they exist
ClearCredentials(host string, path string, username string) error
```
This way, when dealing with repositories which need authentication one can call `CacheCredendials` once with an appropriate timeout and then call other methods of the `Gitter` interface without having to think about the credentials. 

As a further step, one can then imagine providing our own credential helper. There is an [API](https://git-scm.com/docs/api-credentials) for that. This custom credential helper could be a standalone program or a `jx` command. It would then retrieve the credentials from the cluster.

This mechanism would also solve the issue around `jx import` adding the API token to the git remote. 

See also https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage

@pmuir @ccojocar WDYT ^^

**Comment by ccojocar** [comment]
> So issue is that the method in the Gitter interface do not deal with authentication.

This is true, and it is quite of an issue. I kind of fixed this in the auth refactoring which didn't got in. You can have a look at https://github.com/ccojocar/jx/blob/71a3a4c054cba76909e299310d97daf89efd39dc/pkg/gits/git_cli.go#L29 

The `GitCLI` gets an `auth.Server` with the auth config when is created. 

>  The credentials never hit the disk. The idea is to add something along this line to Gitter

I would delegate this to the implementer of Gitter, in this case GitCLI. 

> // CacheCredendials caches the credentials for the specified user, host and path for the specified number of seconds.
CacheCredendials(host string, path string, username string, password string, time int) error
// Clears potentially cached credentials for the specified host, path and user, if they exist
ClearCredentials(host string, path string, username string) error

I am not sure about exposing the credentials like this in the interface. There is already auth.Server which can store the credentials in memory. I would keep them encapsulated in the Gitter implementation because they are sort of related to that. Normally, any helper should delegate to the Gitter implementation for lower level git operations which eventually require a form of authentication.

> As a further step, one can then imagine providing our own credential helper. There is an API for that. This custom credential helper could be a standalone program or a jx command. It would then retrieve the credentials from the cluster.

There are already `jx create git token|server|user` commands which does not work quite right because the persistence logic is mixed with the memory store. The auth.Config/Server/User. (https://github.com/jenkins-x/jx/tree/master/pkg/auth) should be the in-memory cache (data model) and it should support various persistence layers. 

The credentials should be retrieved either from k8s secrets/vault/local files. I think, a lot of the current issues are coming from mixing k8s secret with local files and merging the sources. I think is better to have a credential source top to bottom and fail if they are not found instead of trying to return some inconsistent defaults. 

I agree we should build an interface in parallel with the existing one (one more time :) ) and try to gradually integrate it.

**Comment by ccojocar** [comment]
Also we have git token and git provider token (for API access) which quite often are the same.
